subsys/debug/etb_trace/etb_trace.c

```diff
@@ -0,0 +1,194 @@
+	       != (ETM_TRCSTATR_PMSTABLE | ETM_TRCSTATR_IDLE)) {
+	}
+
+	/* Configure the ETM */
+	SET_REG(ETM_TRCCONFIGR, etm_trcconfigr);
+
+	/* Disable stalling of CPU as the trace unit can not stall the processor
+	 * for instruction traces, at the risk of losing traces.
+	 */
+	SET_REG(ETM_TRCSTALLCTLR, 0);
+
+	/* Set Global Timestamp Control Register to zero so that no stamps included in the trace. */
+	SET_REG(ETM_TRCTSCTLR, 0);
+
+	/* Set the trace stream ID */
+	SET_REG(ETM_TRCTRACEIDR, etm_trctraceidr);
+
+	/* Bit 0: Enable event 0.
+	 * Bit 9: Indicates the current status of the start/stop logic.
+	 * Bit 10: Always trace reset exceptions.
+	 * Bit 11: Always trace system error exceptions.
+	 */
+	SET_REG(ETM_TRCVICTLR,
+		ETM_TRCVICTLR_TRCERR |
+		ETM_TRCVICTLR_TRCRESET |
+		ETM_TRCVICTLR_SSSTATUS |
+		ETM_TRCVICTLR_EVENT0);
+
+	/* No events are configured */
+	SET_REG(ETM_TRCEVENTCTL0R, 0);
+	SET_REG(ETM_TRCEVENTCTL1R, 0);
+
+	/* Enable ETM */
+	SET_REG(ETM_TRCPRGCTLR, ETM_TRCPRGCTLR_ENABLE);
+}
+
+static void etm_stop(void)
+{
+	/* Disable ETM */
+	SET_REG(ETM_TRCPRGCTLR, 0);
+}
+
+static void atb_init(void)
+{
+	/* ATB replicator */
+	CS_UNLOCK(ATB_REPLICATOR_BASE_ADDR);
+
+	/* ID filter for master port 0 */
+	SET_REG(ATB_REPLICATOR_IDFILTER0, 0xFFFFFFFFUL);
+	/* ID filter for master port 1, allowing ETM traces from CM33 to ETB */
+	SET_REG(ATB_REPLICATOR_IDFILTER1, 0xFFFFFFFDUL);
+
+	CS_LOCK(ATB_REPLICATOR_BASE_ADDR);
+
+	/* ATB funnel 1 */
+	CS_UNLOCK(ATB_1_BASE_ADDR);
+
+	/* Set priority 1 for ports 0 and 1 */
+	SET_REG(ATB_1_PRIO, 0x00000009UL);
+
+	/* Enable port 0 and 1, and set hold time to 4 transactions */
+	SET_REG(ATB_1_CTL, 0x00000303UL);
+
+	CS_LOCK(ATB_1_BASE_ADDR);
+
+	/* ATB funnel 2 */
+	CS_UNLOCK(ATB_2_BASE_ADDR);
+
+	/* Set priority 3 for port 3 */
+	SET_REG(ATB_2_PRIO, 0x00003000UL);
+
+	/* Enable ETM traces on port 3, and set hold time to 4 transactions */
+	SET_REG(ATB_2_CTL,  0x00000308UL);
+
+	CS_LOCK(ATB_2_BASE_ADDR);
+}
+
+static void etb_init(void)
+{
+	CS_UNLOCK(ETB_BASE_ADDR);
+
+	/* Disable ETB */
+	SET_REG(ETB_CTL, 0);
+
+	/* Wait for ETB formatter to stop */
+	while (!ETB_FORMATTER_STOPPED) {
+	}
+
+	/* Set ETB formatter to continuous mode */
+	SET_REG(ETB_FFCR, ETB_FFCR_ENFCONT | ETB_FFCR_ENFTC);
+
+	/* Enable ETB */
+	SET_REG(ETB_CTL, ETB_CTL_TRACECAPTEN);
+
+	/* Wait until the ETB Formatter has started */
+	while (ETB_FORMATTER_STOPPED) {
+	}
+
+	CS_LOCK(ETB_BASE_ADDR);
+}
+
+static void etb_stop(void)
+{
+	CS_UNLOCK(ETB_BASE_ADDR);
+
+	/* Disable ETB */
+	SET_REG(ETB_CTL, 0);
+
+	/* Wait for ETB formatter to stop */
+	while (!ETB_FORMATTER_STOPPED) {
+	}
+
+	CS_LOCK(ETB_BASE_ADDR);
+}
+
+static void debug_init(void)
+{
+	NRF_TAD_S->TASKS_CLOCKSTART = TAD_TASKS_CLOCKSTART_TASKS_CLOCKSTART_Msk;
+}
+
+void etb_trace_start(void)
+{
+	debug_init();
+	atb_init();
+	etb_init();
+	etm_init();
+}
+
+void etb_trace_stop(void)
+{
+	etm_stop();
```

Comment:
No need to stop atb?

Reply:
not for power saving - I think ATB just sleeps if there are no messages from the ETM or ITM.